### PR TITLE
[Lambda Extension] Fix decrypting API key with KMS, remove references to SSM

### PR DESCRIPTION
**Pull request by nhinsch**
### What does this PR do?

1. **Fix reading the API key from KMS**
The Lambda console UI changed the way it encrypts environment variables. The current behavior is to encrypt environment variables using the function name as an encryption context. Previously, the behavior was to encrypt environment variables without an encryption context. The Lambda Extension currently only supports API keys encrypted using the *previous* behavior. We need to support both, as supplying the incorrect encryption context will cause decryption to fail.
2. **Remove references to SSM**
The code for retrieving the API key from the AWS Secrets Manager incorrectly referred to it as SSM, which is something else. This PR replaces references to SSM with Secrets Manager, but the functionality is unchanged.

### Describe how to test your changes

I performed manual testing, and also added unit testing for KMS decryption, using a mocked KMS client.

### CODE REVIEW

File: cmd/serverless/api_key.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/kms"
+	"github.com/aws/aws-sdk-go/service/kms/kmsiface"
+	"github.com/aws/aws-sdk-go/service/secretsmanager"
+)
+
+// functionNameEnvVar is the environment variable that stores the Lambda function name
+const functionNameEnvVar = "AWS_LAMBDA_FUNCTION_NAME"
```
Comment (maxday) [inline cmd/serverless/api_key.go:21]: this constant is already here in protocol.go, maybe we should reuse it ?

File: cmd/serverless/api_key.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+// readAPIKeyFromKMS gets and decrypts an API key encrypted with KMS if the env var DD_KMS_API_KEY has been set.
+// If none has been set, it returns an empty string and a nil error.
+func readAPIKeyFromKMS() (string, error) {
+	cipherText := os.Getenv(kmsAPIKeyEnvVar)
+	if cipherText == "" {
+		return "", nil
+	}
+	log.Debug("Found DD_KMS_API_KEY value, trying to decipher it.")
```
Comment (maxday) [inline cmd/serverless/api_key.go:72]: Nit : reuse the `kmsAPIKeyEnvVar` in the log

File: cmd/serverless/api_key.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+// readAPIKeyFromSecretsManager reads an API Key from AWS Secrets Manager if the env var DD_API_KEY_SECRET_ARN has been set.
+// If none has been set, it returns an empty string and a nil error.
+func readAPIKeyFromSecretsManager() (string, error) {
+	arn := os.Getenv(secretsManagerAPIKeyEnvVar)
+	if arn == "" {
+		return "", nil
+	}
+	log.Debug("Found DD_API_KEY_SECRET_ARN value, trying to use it.")
```
Comment (maxday) [inline cmd/serverless/api_key.go:88]: same here

File: cmd/serverless/api_key_test.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
```
Comment (maxday) [inline cmd/serverless/api_key_test.go:1]: 💯

File: cmd/serverless/main.go [hunk 5]
```diff
@@ -180,13 +171,17 @@ func runAgent(stopCh chan struct{}) (daemon *serverless.Daemon, err error) {
...
 	// ---------------
 
+	// API key reading priority:
+	// KSM > Secrets Manager > Plaintext API key
+	// If one is set but failing, the next will be tried
+
 	// some useful warnings first
 
 	var apikeySetIn = []string{}
 	if os.Getenv(kmsAPIKeyEnvVar) != "" {
```
Comment (maxday) [inline cmd/serverless/main.go:181]: we read twice this env variable (once in api_key.go + once here) not sure about how performance is affected but we may want to read it only once if possible ? What do you think ?

File: cmd/serverless/main.go [hunk 6]
```diff
@@ -180,13 +171,17 @@ func runAgent(stopCh chan struct{}) (daemon *serverless.Daemon, err error) {
...
 		apikeySetIn = append(apikeySetIn, "KMS")
 	}
-	if os.Getenv(ssmAPIKeyEnvVar) != "" {
+	if os.Getenv(secretsManagerAPIKeyEnvVar) != "" {
```
Comment (maxday) [inline cmd/serverless/main.go:184]: same here

File: cmd/serverless/main.go [hunk 7]
```diff
@@ -197,24 +192,25 @@ func runAgent(stopCh chan struct{}) (daemon *serverless.Daemon, err error) {
...
-	// try to read apikey from KMS
+	// try to read API key from KMS
 
 	var apiKey string
 	if apiKey, err = readAPIKeyFromKMS(); err != nil {
 		log.Errorf("Error while trying to read an API Key from KMS: %s", err)
 	} else if apiKey != "" {
 		log.Info("Using deciphered KMS API Key.")
-		os.Setenv(apiKeyEnvVar, apiKey) // it will be catched up by config.Load()
+		log.Debug(apiKey)
```
Comment (maxday) [inline cmd/serverless/main.go:202]: This is the plainText API right ? Do we really want to output it ? It may be a security risk ?

Comment (nhinsch) [reply cmd/serverless/main.go:202]: Good catch, this was for my own debugging! Will remove. Although, worth noting that CloudWatch will automatically mask it.

File: cmd/serverless/main.go (hunk 5 above)
Comment (nhinsch) [reply cmd/serverless/main.go:181]: Personally I'm not too worried about it -- this only happens once (on cold start) and I doubt the penalty for reading an environment variable would be noticeable/worth optimizing at this point.

File: cmd/serverless/main.go (hunk 6 above)
Comment (nhinsch) [reply cmd/serverless/main.go:184]: see other comment

**Review by maxday — APPROVED** [review]

**Pull request merged**